### PR TITLE
Add printing system to RichDocuments

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
- open up a document with Collabora
- select File -> print
- printable pdf file will be downloaded and sent to print system

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
OK, while most Codacy issues should be resolved 👍

File: src/main/java/com/owncloud/android/ui/asynctasks/PrintAsyncTask.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            if (status == 200) {
+                if (file.exists()) {
+                    file.delete();
+                }
+
+                file.getParentFile().mkdirs();
+
+                file.createNewFile();
+                BufferedInputStream bis = new BufferedInputStream(mGet.getResponseBodyAsStream());
+                fos = new FileOutputStream(file);
```
[the remark this answers is not in the document]
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/asynctasks/PrintAsyncTask.java]: We cannot do this as this requires API >= 26.

File: src/main/java/com/owncloud/android/ui/activity/RichDocumentsWebView.java [hunk 2]
```diff
@@ -384,27 +422,22 @@ public void documentLoaded() {
...
             try {
                 JSONObject downloadJson = new JSONObject(json);
-                downloadUrl = Uri.parse(downloadJson.getString("URL"));
-            } catch (JSONException e) {
-                Log_OC.e(this, "Failed to parse rename json message: " + e);
-                return;
-            }
 
-            DownloadManager downloadmanager = (DownloadManager) getSystemService(Context.DOWNLOAD_SERVICE);
+                Uri url = Uri.parse(downloadJson.getString("URL"));
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/activity/RichDocumentsWebView.java]: "URL" should be a constant field

File: src/main/java/com/owncloud/android/ui/activity/RichDocumentsWebView.java [hunk 3]
```diff
@@ -384,27 +422,22 @@ public void documentLoaded() {
...
 
-            if (downloadmanager == null) {
-                DisplayUtils.showSnackMessage(webview, getString(R.string.failed_to_download));
+                if (downloadJson.getString("Type").equalsIgnoreCase("print")) {
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/activity/RichDocumentsWebView.java]: "Type" and "print" should be a constant fields

File: src/main/java/com/owncloud/android/ui/adapter/PrintAdapter.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    @Override
+    public void onLayout(PrintAttributes oldAttributes,
+                         PrintAttributes newAttributes,
+                         CancellationSignal cancellationSignal,
+                         LayoutResultCallback callback,
+                         Bundle extras) {
+        if (cancellationSignal.isCanceled()) {
+            callback.onLayoutCancelled();
+        } else {
+            PrintDocumentInfo.Builder builder = new PrintDocumentInfo.Builder("finalPrint.pdf");
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/adapter/PrintAdapter.java]: "finalPrint.pdf" should be a constant field

File: src/main/java/com/owncloud/android/ui/asynctasks/PrintAsyncTask.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    public PrintAsyncTask(File file, String url, WeakReference<RichDocumentsWebView> richDocumentsWebViewWeakReference) {
+        this.file = file;
+        this.url = url;
+        this.richDocumentsWebViewWeakReference = richDocumentsWebViewWeakReference;
+    }
+
+    @Override
+    protected void onPreExecute() {
+        richDocumentsWebViewWeakReference.get().runOnUiThread(
+            () -> richDocumentsWebViewWeakReference.get().showLoadingDialog("Loading…"));
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/asynctasks/PrintAsyncTask.java]: "Loading…" should be translatable.

File: src/main/java/com/owncloud/android/ui/asynctasks/PrintAsyncTask.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        if (!result || printManager == null) {
+            DisplayUtils.showSnackMessage(richDocumentsWebView,
+                                          richDocumentsWebView.getString(R.string.failed_to_print));
+
+            return;
+        }
+
+        PrintDocumentAdapter printAdapter = new PrintAdapter(file.getAbsolutePath());
+
+        printManager.print("Document", printAdapter, new PrintAttributes.Builder().build());
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/asynctasks/PrintAsyncTask.java]: "Document" should be a constant field

**Comment by tobiasKaminsky** [comment]
I increase spotbugs as  this is the same Codacy found, but for all StreamWriter we cannot switch as we still have to support old APIs.

**Pull request merged**